Flashforge Adventurer 4, 5M, 5M Pro support

f00d4tehg0dz opened the pull request:
- Added support for Adventurer 4, 5M and 5M Pro
- Added selector system for configuration to choose your Flashforge Added regex for additional Adventurer 4 attributes 
- Added support for Flashforge Finder. Same regex as Adventurer 3

-- **Need Testing** --
- Flashforge Adventurer 5M
- Flashforge Adventurer 4

**Screenshots**

![configuration-card](https://github.com/modrzew/hass-flashforge-adventurer-3/assets/1635526/8c477ace-9718-452d-9cec-3ee87960d1aa)

![card](https://github.com/modrzew/hass-flashforge-adventurer-3/assets/1635526/12b88134-a95e-4832-a769-bb177cda04d5)

![nozzle-attributes](https://github.com/modrzew/hass-flashforge-adventurer-3/assets/1635526/e5faef88-9307-4c45-801a-438951bd45bd)

![progress-attributes](https://github.com/modrzew/hass-flashforge-adventurer-3/assets/1635526/b0a7821e-caaa-4813-8469-90e8e06a63fa)

![state-attributes](https://github.com/modrzew/hass-flashforge-adventurer-3/assets/1635526/23e51c7b-e133-49c0-aa7c-cf567c77a019)

f00d4tehg0dz committed:
Flashforge Adventurer 5M, 5M Pro

Added supported for Adventurer 5M, 5M Pro
Added selector system for configuration to choose your Flashforge
Added regex for additional Adventurer 4 attributes support
Added Flashforge original Finder. Uses same regex as Adventurer 3

-- Need Testing --
- Flashforge Adventurer 5M
- Flashforge Adventurer 4

gelimelech commented:
How do i install this update? 🙏🏾

f00d4tehg0dz commented:
> How do i install this update? 🙏🏾

The normal method won't be available as @modrzew needs to review, ask for changes if necessary and approve the PR, and merge.

For now you can simple download the .zip here: 
- https://github.com/f00d4tehg0dz/hass-flashforge-adventurer-3/archive/refs/heads/feature/flashforge-adventurer-5m-pro.zip
- Extract the `flashforge_adventurer_3` folder (which is inside the `custom_components` folder)
- Copy the folder to your Home Assistant HACS `custom_components` directory. That could look like /config/home-assistant/custom_components/flashforge_adventurer_3
- Restart Home Assistant
- Add Integration
- Add Flashforge Adventurer 3
- Input your 3D Printers IP and select the model you have
- Add the respective cards to your Home Assistant Dashboard 

My card YAML for reference. I'm using the `mushroom-entity-card` template however.
```
type: vertical-stack
cards:
  - show_state: false
    show_name: false
    camera_view: auto
    type: picture-entity
    entity: camera.flashforge_adventurer_5m_pro_camera
    camera_image: camera.flashforge_adventurer_5m_pro_camera
    name: Camera
    hold_action:
      action: more-info
  - type: gauge
    entity: sensor.flashforge_adventurer_5m_pro_progress
    name: Progress
  - type: custom:mushroom-entity-card
    entity: sensor.flashforge_adventurer_5m_pro_state
    name: State
  - type: entities
    entities:
      - entity: sensor.flashforge_adventurer_5m_pro_nozzle_temperature
        secondary_info: none
        name: Nozzle
    show_header_toggle: false

```

gelimelech commented:
How do i install this update? 🙏🏾

> > How do i install this update? 🙏🏾
> 
> The normal method won't be available as @modrzew needs to review, ask for changes if necessary and approve the PR, and merge.
> 
> For now you can simple download the .zip here:
> 
>     * https://github.com/f00d4tehg0dz/hass-flashforge-adventurer-3/archive/refs/heads/feature/flashforge-adventurer-5m-pro.zip
> 
>     * Extract the `flashforge_adventurer_3` folder (which is inside the `custom_components` folder)
> 
>     * Copy the folder to your Home Assistant HACS `custom_components` directory. That could look like /config/home-assistant/custom_components/flashforge_adventurer_3
> 
>     * Restart Home Assistant
> 
>     * Add Integration
> 
>     * Add Flashforge Adventurer 3
> 
>     * Input your 3D Printers IP and select the model you have
> 
>     * Add the respective cards to your Home Assistant Dashboard
> 
> 
> My card YAML for reference. I'm using the `mushroom-entity-card` template however.
> 
> ```
> type: vertical-stack
> cards:
>   - show_state: false
>     show_name: false
>     camera_view: auto
>     type: picture-entity
>     entity: camera.flashforge_adventurer_5m_pro_camera
>     camera_image: camera.flashforge_adventurer_5m_pro_camera
>     name: Camera
>     hold_action:
>       action: more-info
>   - type: gauge
>     entity: sensor.flashforge_adventurer_5m_pro_progress
>     name: Progress
>   - type: custom:mushroom-entity-card
>     entity: sensor.flashforge_adventurer_5m_pro_state
>     name: State
>   - type: entities
>     entities:
>       - entity: sensor.flashforge_adventurer_5m_pro_nozzle_temperature
>         secondary_info: none
>         name: Nozzle
>     show_header_toggle: false
> ```

Thank you very much!
it wotks 😁

mistraller commented:
> * Added support for Adventurer 4, 5M and 5M Pro
> * Added selector system for configuration to choose your Flashforge Added regex for additional Adventurer 4 attributes
> * Added support for Flashforge Finder. Same regex as Adventurer 3
> 
> -- **Need Testing** --
> 
> * Flashforge Adventurer 5M
> * Flashforge Adventurer 4
> 
Running your code. Adventurer 5M pro. All seems fine. I had to figure out how to show the attributes in Lovelace though. But it works.

mistraller commented:
@f00d4tehg0dz 
Saw these in logs. (Printer is powered off btw)
Logger: homeassistant.helpers.integration_platform
Bron: helpers/integration_platform.py:52
First occurred: 00:24:38 (9 gebeurtenissen)
Laatst gelogd: 00:26:44

Unexpected error importing FlashForge_adventurer_3/logbook.py
Unexpected error importing FlashForge_adventurer_3/system_health.py
Unexpected error importing FlashForge_adventurer_3/cast.py
Unexpected error importing FlashForge_adventurer_3/energy.py
Unexpected error importing FlashForge_adventurer_3/hardware.py
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/integration_platform.py", line 52, in _get_platform
    return integration.get_platform(platform_name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/loader.py", line 1010, in get_platform
    return self._load_platform(platform_name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/loader.py", line 1025, in _load_platform
    cache[full_name] = self._import_platform(platform_name)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/loader.py", line 1058, in _import_platform
    return importlib.import_module(f"{self.pkg_path}.{platform_name}")
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/importlib/__init__.py", line 90, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1324, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'custom_components.flashforge_adventurer_3.diagnostics'

modrzew commented on custom_components/flashforge_adventurer_3/const.py:
```diff
@@ -1,3 +1,14 @@
+
+# Add constants for printer models
+CONF_MODEL = 'model'
+MODEL_FINDER = 'FlashForge Finder'
+MODEL_ADVENTURER_3 = 'FlashForge Adventurer 3'
+MODEL_ADVENTURER_4 = 'FlashForge Adventurer 4'
+MODEL_ADVENTURER_5M = 'FlashForge Adventurer 5M'
+MODEL_ADVENTURER_5M_PRO = 'FlashForge Adventurer 5M Pro'
+AVAILABLE_MODELS = [MODEL_FINDER, MODEL_ADVENTURER_3, MODEL_ADVENTURER_4, MODEL_ADVENTURER_5M, MODEL_ADVENTURER_5M_PRO]
```
At the moment the human-readable name of the printer model is the same as its identifier used in the code, for example to choose the most appropriate regexp. Let's separate these two concerns.

Firstly, let's use shorter identifiers throughout the codebase, as well as an identifier persisted in the integration:

- `finder`
- `adventurer_3`
- `adventurer_4`
- `adventurer_5m`
- `adventurer_5m_pro

Up to you if you want to export these as constants and use those constants, or if you want to use string literals in `protocol.py`.

Secondly, let's create a map of those identifiers to human-readable printer names that will be used in the UI, with values being the identifiers above and labels being the strings that you have now.

modrzew commented on custom_components/flashforge_adventurer_3/protocol.py:
```diff
@@ -1,30 +1,38 @@
+
+# Regular expressions for temperature and progress
+
+TEMPERATURE_REGEX_5M_PRO = re.compile(r'T0:(\d+\.?\d*)/(\d+\.?\d*) .* B:(\d+\.?\d*)/(\d+\.?\d*)')
+PROGRESS_REGEX_5M_PRO = re.compile(r'(\d+)/(\d+)\r')
```
Since it's for both 5M and 5M Pro, let's use the `_5M` suffix (rather than `_5M_PRO`).

modrzew commented on custom_components/flashforge_adventurer_3/protocol.py:
```diff
@@ -1,30 +1,38 @@
+TEMPERATURE_REGEX_5M_PRO = re.compile(r'T0:(\d+\.?\d*)/(\d+\.?\d*) .* B:(\d+\.?\d*)/(\d+\.?\d*)')
+PROGRESS_REGEX_5M_PRO = re.compile(r'(\d+)/(\d+)\r')
+
+TEMPERATURE_REGEX_ADVENTURER_4 = re.compile(r'CMD M105 Received.\r\nT0:(\d+)\W*/(\d+) B:(\d+)\W*/(\d+)\r\n(.*?)ok\r\n')
```
Temperature regexes for 4 and 3 are the same. Let's use the same constant for both of them rather than duplicating.

modrzew commented on custom_components/flashforge_adventurer_3/protocol.py:
```diff
@@ -1,30 +1,38 @@
-STATUS_COMMAND = '~M601 S1'
-PRINT_JOB_INFO_COMMAND = '~M27'
-TEMPERATURE_COMMAND = '~M105'
+STATUS_COMMAND = '~M119\r\n'
```
Looks like the status command is no longer used. I guess we can remove it? (I was never sure if it was needed in the first place 😛)

modrzew commented on custom_components/flashforge_adventurer_3/protocol.py:
```diff
@@ -35,48 +43,54 @@ async def send_msg(reader: StreamReader, writer: StreamWriter, payload: str):
 
-
-async def collect_data(ip: str, port: int) -> Tuple[PrinterStatus, Optional[str], Optional[str]]:
+async def collect_data(ip: str, port: int, model: str) -> Tuple[PrinterStatus, Optional[str], Optional[str]]:
```
Let's use strong typing for `model` rather than passing in any arbitrary string. Either a string literal or an enum can work.

modrzew commented on custom_components/flashforge_adventurer_3/protocol.py:
```diff
@@ -1,30 +1,38 @@
-TEMPERATURE_REPLY_REGEX = re.compile('CMD M105 Received.\r\nT0:(\d+)\W*/(\d+) B:(\d+)\W*/(\d+)\r\n(.*?)ok\r\n')
 
 class PrinterStatus(TypedDict):
+    model: str
```
Let's use strong typing for `model` rather than passing in any arbitrary string. Either a string literal or an enum can work.

modrzew commented on custom_components/flashforge_adventurer_3/protocol.py:
```diff
@@ -35,48 +43,54 @@ async def send_msg(reader: StreamReader, writer: StreamWriter, payload: str):
 
+async def get_print_job_status(ip: str, port: int, model: str) -> PrinterStatus:
+    response, print_job_info, temperature_info = await collect_data(ip, port, model)
+    if response['online']:
+        return parse_data(model, print_job_info, temperature_info)
```
Why did you invert the conditional?

modrzew commented on custom_components/flashforge_adventurer_3/sensor.py:
```diff
@@ -14,74 +14,91 @@
-        vol.Required('ip'): cv.string,
+        vol.Required('ip_address'): cv.string,
         vol.Required('port'): cv.string,
+        vol.Required(CONF_MODEL): cv.string,
```
Can we use `cv.enum` here?

modrzew commented on custom_components/flashforge_adventurer_3/sensor.py:
```diff
@@ -14,74 +14,91 @@
 ) -> bool:
+
+    logger = logging.getLogger(__name__)
+    logger.info("Setting up sensor platform for entry %s", config_entry.entry_id)
```
Can all of these logging statements be `debug` rather than `info`?

modrzew commented on custom_components/flashforge_adventurer_3/sensor.py:
```diff
@@ -152,3 +169,32 @@ def icon(self) -> str:
     def unit_of_measurement(self) -> str:
         return '%'
+
+class FlashForgeAdventurer5TemperatureSensor(BaseFlashForgeAdventurer3Sensor):
```
Nozzle and bed temperatures are kept as attributes on the state sensor, and can be accessed via it. I don't think this new sensor is necessary to be part of the integration.

Geoffery10 replied:
I thought it was somewhat odd in my testing that the nozzle temperature was the only one with its own sensor. 🤔

f00d4tehg0dz replied:
Got it. I'll remove it. :)

modrzew commented on custom_components/flashforge_adventurer_3/strings.json:
```diff
@@ -3,12 +3,13 @@
-        "title": "FlashForge Adventurer 3 setup",
+        "title": "FlashForge Printer Setup",
         "data": {
+          "model": "Printer Model (Adventurer 3, Adventurer 5M Pro, etc.)",
```
```suggestion
          "model": "Printer model",
```

given that it'll display all supported models underneath.

modrzew commented on custom_components/flashforge_adventurer_3/sensor.py:
```diff
@@ -14,74 +14,91 @@
         )
-        self.ip = printer_definition['ip_address']
-        self.port = printer_definition['port']
+        self.ip = printer_definition.get('ip_address', 'Unknown IP')  # Set a default IP if not provided
+        self.port = printer_definition.get('port', 8899)  # Set a default port if not provided
+        self.model = printer_definition.get(CONF_MODEL, 'FlashForge Adventurer 3')
+
```
🤔 I think it'll be better for the integration to fail early, rather than try to telnet to `Unknown IP`.

modrzew commented on custom_components/flashforge_adventurer_3/sensor.py:
```diff
@@ -14,74 +14,91 @@
     def __init__(self, coordinator: DataUpdateCoordinator, printer_definition: PrinterDefinition) -> None:
         super().__init__(coordinator)
-        self.ip = printer_definition['ip_address']
+        self.ip = printer_definition['ip_address']  # Use 'ip_address' instead of 'ip'
```
What does this comment mean?

f00d4tehg0dz replied:
That's my mistake. I had written it as `['ip']` and left a comment for myself to not make a dumb mistake. I'll remove

modrzew reviewed (changes requested):
Hey, sorry for the huge delay and thank you for opening this PR!

Can you tell me which printers you've tested these changes on? Sadly I no longer have my Adventurer 3, so we'll need to work around that with testing 😅 

Also, two nits:
- `Flashforge` is intentionally spelled this way in all the class names, rather than `FlashForge`. Can you restore the prior naming? It'll also reduce the number of affected lines in this PR making it easier to review the changes.
- Can you leave new line symbols at the end of files rather than removing them?

Geoffery10 commented:
I tested it on my Flashforge Adventure 5 and it works without issue. 👍

f00d4tehg0dz commented:
> Can you tell me which printers you've tested these changes on? Sadly I no longer have my Adventurer 3, so we'll need to work around that with testing 😅

Hey! No problem. Unfortunately, I don't have Adventurer 3 either. I have Flashforge Finder and Flashforge 5M Pro. I made an assumption that the Flashforge 5M would be the same as Flashforge 5M Pro. However, I'm happy to remove that model as it's not best practice to assume.
Based on other comments, Adventurer 4 worked as well. I was debating making a specific exception for Flashforge Finder to disable the camera as that model never had one 😉

I will go through and revert some of the changes and adjust accordingly based on your comments. Thanks so much for replying back! ❤️

modrzew commented:
As I don't have my Adventurer 3 anymore, it's hard for me to verify the changes anymore anyway. So maybe let's leave the 5M under the assumption that it works simialrly to 5M Pro, and in case something goes wrong we can always fix it later 🙂

mistraller commented:
> As I don't have my Adventurer 3 anymore, it's hard for me to verify the changes anymore anyway. So maybe let's leave the 5M under the assumption that it works simialrly to 5M Pro, and in case something goes wrong we can always fix it later 🙂

Main difference between 5m and 5mpro is camera.
Did you test already with the new flashforge firmware? There are many complaints about camera on 5m pro after installing the new 2.6.5 firmware.

f00d4tehg0dz commented:
> > As I don't have my Adventurer 3 anymore, it's hard for me to verify the changes anymore anyway. So maybe let's leave the 5M under the assumption that it works simialrly to 5M Pro, and in case something goes wrong we can always fix it later 🙂
> 
> Main difference between 5m and 5mpro is camera. Did you test already with the new flashforge firmware? There are many complaints about camera on 5m pro after installing the new 2.6.5 firmware.

No issues with the camera for 5M Pro. I just installed the new firmware and tested. Works just fine.  

I plan on having all the changes submitted by EOW. I'll keep you all posted. Thanks!

timnolte commented:
I have just received a 5M, not Pro, so I'd be happy to test and confirm it if needed. Looking forward to having this integration available.

f00d4tehg0dz committed:
requested edits

* Revert unnecessary changes, such as Flashforge to FlashForge and \r\n
* Logger uses debug
* Removed dedicated Flashforge Adventurer 5 Temp Sensor
* Created map of identifiers to human readable printer names
* Incorporated map throughout files
* Use _5M  suffix
* Removed additional temperature regex when adventurer 3 and 4 are the same
* Changed model: str with using PrinterModel
* Used enum

f00d4tehg0dz reviewed (commented):
I believe I have addressed all your comments. Take your time to review! Thanks :)

f00d4tehg0dz commented:
> I have just received a 5M, not Pro, so I'd be happy to test and confirm it if needed. Looking forward to having this integration available.

I'd love for you to test. I just submitted changes in this PR. Feel free to grab the PR and test it out on your HA.

samjones108 commented:
Just a ping regarding the adventurer 4, 

Progress bar doesn't update when printing stays on 0%.
Camera works fine no issues there.
state works fine changes between printing/idle. 

Not sure if you can pull temperature status on the adventure 4 as that's not showing either.

samjones108 commented:
Unsure, Just found the addon today. the progress was working with the published version using the adventurer 3 settings but not on the above zip version.

f00d4tehg0dz commented:
> Unsure, Just found the addon today. the progress was working with the published version using the adventurer 3 settings but not on the above zip version.

Got it. Thanks for clarifying. Let me see if I made a mistake that affected you somewhere.

f00d4tehg0dz committed:
Dedicated Adventurer 5 status reply, restore to original authors classnames

* PROGRESS_REGEX_ is back to STATUS_REPLY_REGEX_
* progress_match is back to print_job_info_match
* Adventurer 4 has it's own section to handle layers and total layers functionality

f00d4tehg0dz commented:
> Unsure, Just found the addon today. the progress was working with the published version using the adventurer 3 settings but not on the above zip version.

Give this version a [shot ](https://github.com/f00d4tehg0dz/hass-flashforge-adventurer-3/archive/refs/tags/v1.0.5.zip) if you'd like. Progress bar should work now. Regarding the temperature not working, I'm not sure what I'm missing there. I went through @modrzew  codebase  and it appeared to be similar to mine. 

https://github.com/f00d4tehg0dz/hass-flashforge-adventurer-3/archive/refs/tags/v1.0.5.zip

f00d4tehg0dz committed:
Increase update frequency & add layers printed and total layers for 5M & 5MP

Increased update frequency from 60 to 30 now that printing layers are supported
5M and 5MP support viewing total layers to print, and how many layers are printed so far

timnolte commented:
@f00d4tehg0dz something seems off. I downloaded the 1.0.6 versions from here: https://github.com/f00d4tehg0dz/hass-flashforge-adventurer-3/archive/refs/tags/v1.0.6.zip but when I installed it and restarted HA I get no prompt for selecting my Adventurer model.

timnolte commented:
> @f00d4tehg0dz something seems off. I downloaded the 1.0.6 versions from here: https://github.com/f00d4tehg0dz/hass-flashforge-adventurer-3/archive/refs/tags/v1.0.6.zip but when I installed it and restarted HA I get no prompt for selecting my Adventurer model.

I'm going to try the v1.0.5.zip to see if it's any different.

f00d4tehg0dz commented:
> > @f00d4tehg0dz something seems off. I downloaded the 1.0.6 versions from here: https://github.com/f00d4tehg0dz/hass-flashforge-adventurer-3/archive/refs/tags/v1.0.6.zip but when I installed it and restarted HA I get no prompt for selecting my Adventurer model.
> 
> I'm going to try the v1.0.5.zip to see if it's any different.

aww, dang, that's concerning... When I'm off work, I'll play around and try a fresh install and see. 

Did you, by chance, restart HA entirely? Or a refresh? ( I realize you said restart, but just in case...)

timnolte commented:
> > > @f00d4tehg0dz something seems off. I downloaded the 1.0.6 versions from here: https://github.com/f00d4tehg0dz/hass-flashforge-adventurer-3/archive/refs/tags/v1.0.6.zip but when I installed it and restarted HA I get no prompt for selecting my Adventurer model.
> > 
> > 
> > I'm going to try the v1.0.5.zip to see if it's any different.
> 
> aww, dang, that's concerning... When I'm off work, I'll play around and try a fresh install and see.
> 
> Did you, by chance, restart HA entirely? Or a refresh? ( I realize you said restart, but just in case...)

So, I'm not sure what's going on, the v1.0.5.zip is also not working. Yes, I did a full HA restart. I never had the integration installed and I can see that it's not present before a restart but is afterwards. I'm going to try the zip archive of your branch instead of the tags. Oh, wait a minute. So yeah, your tags in your repo both v1.0.5 and v1.0.6 are of the same commit.

![image](https://github.com/modrzew/hass-flashforge-adventurer-3/assets/4312062/b1b366d5-ba09-49e9-b3c7-5102b3b2bc34)

And looking at the source of that commit and you aren't actually tagging your commits.

![image](https://github.com/modrzew/hass-flashforge-adventurer-3/assets/4312062/b06b1696-47a6-40ac-bedf-9eb20131ea0d)

f00d4tehg0dz commented:
> > > > @f00d4tehg0dz something seems off. I downloaded the 1.0.6 versions from here: https://github.com/f00d4tehg0dz/hass-flashforge-adventurer-3/archive/refs/tags/v1.0.6.zip but when I installed it and restarted HA I get no prompt for selecting my Adventurer model.
> > > 
> > > 
> > > I'm going to try the v1.0.5.zip to see if it's any different.
> > 
> > 
> > aww, dang, that's concerning... When I'm off work, I'll play around and try a fresh install and see.
> > Did you, by chance, restart HA entirely? Or a refresh? ( I realize you said restart, but just in case...)
> 
> So, I'm not sure what's going on, the v1.0.5.zip is also not working. Yes, I did a full HA restart. I never had the integration installed and I can see that it's not present before a restart but is afterwards. I'm going to try the zip archive of your branch instead of the tags. Oh, wait a minute. So yeah, your tags in your repo both v1.0.5 and v1.0.6 are of the same commit.
> 
> ![image](https://private-user-images.githubusercontent.com/4312062/334554215-b1b366d5-ba09-49e9-b3c7-5102b3b2bc34.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY5MjU4MjQsIm5iZiI6MTcxNjkyNTUyNCwicGF0aCI6Ii80MzEyMDYyLzMzNDU1NDIxNS1iMWIzNjZkNS1iYTA5LTQ5ZTktYjNjNy01MTAyYjNiMmJjMzQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MjhUMTk0NTI0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Y2ExZmRmODI3NjVhMWMzYmUxMTVmNzcxMzdmMGQ5MDgzOWQ3MDMyZmY1NWFkNDUyN2I1YzQ2ZDdlYjlmZjc1YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.FWnhn6tkCWGJrRSrjj3TPnaGVGEw01v9JLEOQYXt01M)
> 
> And looking at the source of that commit and you aren't actually tagging your commits.
> 
> ![image](https://private-user-images.githubusercontent.com/4312062/334554450-b06b1696-47a6-40ac-bedf-9eb20131ea0d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY5MjU4MjQsIm5iZiI6MTcxNjkyNTUyNCwicGF0aCI6Ii80MzEyMDYyLzMzNDU1NDQ1MC1iMDZiMTY5Ni00N2E2LTQwYWMtYmVkZi05ZWIyMDEzMWVhMGQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MjhUMTk0NTI0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDM1YThlYTJjM2Q0YzY3ZTA0ODc4Yzc4OGQ0NzQyYzIwODE5NDQwZThkY2RiZmQxNDhjY2U0NmM0MmY5N2NiYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.49hwJ0eKAnPyzrSEQcb30V2yxGxVpExbKq3eLkFZZoI)

Holy cow. You're right. https://github.com/f00d4tehg0dz/hass-flashforge-adventurer-3/releases/tag/v1.0.7 Now it's tagged correctly. Yikes. Thank you for catching this!

mistraller commented:
Thinking about layers-to-print and layers done; is that not just the way how the progress is calculated? 
My AD5M shows it progress the same way; if the last layers are less work (and that is most of the time unless I'm printing a cube ;) ), the precentage will go much quicker.

timnolte commented:
So I just happened to take a look at the state when my print job was finished and there are a bunch of attributes that are already available. If you want to display them you could easily create some custom sensors to display those values in a dedicated way.

![Screenshot_20240528-181743.png](https://github.com/modrzew/hass-flashforge-adventurer-3/assets/4312062/8ac74409-264d-4dd7-b7f1-becc21de9f58)

timnolte commented:
Something that is interesting with those values is that the print job is complete at 100% but it says it printing layer 401 with all layers being 411. 🤔

f00d4tehg0dz commented:
> So I just happened to take a look at the state when my print job was finished and there are a bunch of attributes that are already available. If you want to display them you could easily create some custom sensors to display those values in a dedicated way.
> 
>
Correct, that is intentional. You can use an Entity card to split out each attribute. I just took a screenshot of a current print. I originally had included one of the attributes as a sensor. However, during the PR review, it was [requested to remove that sensor](https://github.com/modrzew/hass-flashforge-adventurer-3/pull/17#discussion_r1563829206) and leave all as attributes. 
![image](https://github.com/modrzew/hass-flashforge-adventurer-3/assets/1635526/517e330a-0465-40a7-9f01-ba2b305adfa2)

f00d4tehg0dz commented:
> Something that is interesting with those values is that the print job is complete at 100% but it says it printing layer 401 with all layers being 411. 🤔

That is interesting. I'll do some debugging on that. I had increased the update [interval to 30 seconds](https://github.com/modrzew/hass-flashforge-adventurer-3/blob/65a91aee81c9c7cbc39e465fa0ba75b05315066f/custom_components/flashforge_adventurer_3/sensor.py#L76) from 60 seconds. However, the layers shouldn't have been off that much...

timnolte commented:
> > Something that is interesting with those values is that the print job is complete at 100% but it says it printing layer 401 with all layers being 411. 🤔
> 
> That is interesting. I'll do some debugging on that. I had increased the update [interval to 30 seconds](https://github.com/modrzew/hass-flashforge-adventurer-3/blob/65a91aee81c9c7cbc39e465fa0ba75b05315066f/custom_components/flashforge_adventurer_3/sensor.py#L76) from 60 seconds. However, the layers shouldn't have been off that much... 

So here is something equally strange. The total layers is printing as the same value as the current layer during printing.

![Screenshot_20240529-191334.png](https://github.com/modrzew/hass-flashforge-adventurer-3/assets/4312062/06a56eee-f748-4a53-9ca4-f2d41ee9cc20)

f00d4tehg0dz commented:
> > > Something that is interesting with those values is that the print job is complete at 100% but it says it printing layer 401 with all layers being 411. 🤔
> > 
> > 
> > That is interesting. I'll do some debugging on that. I had increased the update [interval to 30 seconds](https://github.com/modrzew/hass-flashforge-adventurer-3/blob/65a91aee81c9c7cbc39e465fa0ba75b05315066f/custom_components/flashforge_adventurer_3/sensor.py#L76) from 60 seconds. However, the layers shouldn't have been off that much...
> 
> So here is something equally strange. The total layers is showing as the same value as the current layer during printing.


I may limit the total layer to Adventurer 4 and 5M Pro, as they appear to be working there. I highly doubt the 5M would be any different, though.

timnolte commented:
> > > > Something that is interesting with those values is that the print job is complete at 100% but it says it printing layer 401 with all layers being 411. 🤔
> > > 
> > > 
> > > That is interesting. I'll do some debugging on that. I had increased the update [interval to 30 seconds](https://github.com/modrzew/hass-flashforge-adventurer-3/blob/65a91aee81c9c7cbc39e465fa0ba75b05315066f/custom_components/flashforge_adventurer_3/sensor.py#L76) from 60 seconds. However, the layers shouldn't have been off that much...
> > 
> > So here is something equally strange. The total layers is showing as the same value as the current layer during printing.
> 
> 
> I may limit the total layer to Adventurer 4 and 5M Pro, as they appear to be working there. I highly doubt the 5M would be any different, though.
> 

To a degree, other than the prebuilt enclosure and hardware (the camera and the air filtration system), I thought that the 5M Pro software was essentially the same as the 5M.

timnolte commented:
I'll be honest that I really feel like this integration should be providing an actual Adventurer device that encompasses the entities and that all of these attributes should actually just be sensors of that device with a small set of them enabled by default with the rest not enabled. Being forced to create all of the sensors manually or setup custom cards to display attributes is a pain. Not to mention that without custom sensors you can't capture graphs of say temperature changes over time or things like the actual printing state over time. I still can't get my HA notifications to work right with this setup to notify me when the printing is finished.

f00d4tehg0dz commented:
> I'll be honest that I really feel like this integration should be providing an actual Adventurer device that encompasses the entities and that all of these attributes should actually just be sensors of that device with a small set of them enabled by default with the rest not enabled. Being forced to create all of the sensors manually or setup custom cards to display attributes is a pain. Not to mention that without custom sensors you can't capture graphs of say temperature changes over time or things like the actual printing state over time. I still can't get my HA notifications to work right with this setup to notify me when the printing is finished.

Noted. And I don't disagree with you. Creating the sensors is certainly doable, I could continue this project with a sensor variant on a separate repo.

f00d4tehg0dz commented:
> gree with you. Creating the sensors is certainly doable, I could continue this project with a sensor variant on a separate repo.

@modrzew, what's your opinion on this? I don't want to create a bunch of sensors if you disagree. 
FWIW, I just installed the Klipper mod to my 5MP (easily reversible) to build out the Rabbit feeder mod (8 or more color filament automation) to support Klipper on HA for 5MP. I'll need to modify your source code to further support a Klipper variant. At which point it might make sense to fork and go on my own path. However, I want to hear your thoughts and concerns!

timnolte commented:
> > gree with you. Creating the sensors is certainly doable, I could continue this project with a sensor variant on a separate repo.
> 
> @modrzew, what's your opinion on this? I don't want to create a bunch of sensors if you disagree. 
> FWIW, I just installed the Klipper mod to my 5MP (easily reversible) to build out the Rabbit feeder mod (8 or more color filament automation) to support Klipper on HA for 5MP. I'll need to modify your source code to further support a Klipper variant. At which point it might make sense to fork and go on my own path. However, I want to hear your thoughts and concerns!

@f00d4tehg0dz & @modrzew just to be clear I'm not advocating the creation of a device and more sensors in this branch. I'd actually say let's push that to a new set of work so as not to delay a baseline of 4 & 5M/5MP support.

f00d4tehg0dz commented:
> > > gree with you. Creating the sensors is certainly doable, I could continue this project with a sensor variant on a separate repo.
> > 
> > 
> > @modrzew, what's your opinion on this? I don't want to create a bunch of sensors if you disagree.
> > FWIW, I just installed the Klipper mod to my 5MP (easily reversible) to build out the Rabbit feeder mod (8 or more color filament automation) to support Klipper on HA for 5MP. I'll need to modify your source code to further support a Klipper variant. At which point it might make sense to fork and go on my own path. However, I want to hear your thoughts and concerns!
> 
> @f00d4tehg0dz & @modrzew just to be clear I'm not advocating the creation of a device and more sensors in this branch. I'd actually say let's push that to a new set of work so as not to delay a baseline of 4 & 5M/5MP support.

Ah sorry. I misunderstood your previous comments on this. Thanks for clarifying!

modrzew commented:
Hey both, thank you for your input so far! I don't mind this integration having more sensors. I didn't have any grand vision when I put things like temperatures in the attributes, rather than creating separate sensors for them — it was purely driven by what I needed in my dashboard at the time, and I simply thought "I don't need to track these, let's chuck them in attributes" 😄 also +1 to doing this work outside of this branch — let's keep this one about supporting more printers.

Moreover, I'm less invested in this integration right now as I parted ways with my Adventurer 3 a few months ago — I'm not using it anymore, and I can't even test the changes.

@f00d4tehg0dz I think it might be best for you to continue the development in your fork, if you're keen. Once you're happy with it we can swap the reference in HACS to point to it rather than to this repo, and I can archive this one. WDYT?

(This might also be an opportunity to rename the repo/integration, since it's no longer just Adventurer 3 🙂)

timnolte commented:
So not sure where things stand. For the most part I'm generally having good success with the Adventurer 5M support. I even setup my own dedicated 3D Printing monitoring dashboard in HA.

![Screenshot_20240602-235707.png](https://github.com/modrzew/hass-flashforge-adventurer-3/assets/4312062/073f210a-d23c-4314-af45-05cd04d38413)

I use a Tapo smart security camera for my monitoring so haven't tested the built-in camera support functionality. I also have print finished notifications working through HA. Since I also have my printer plugged into a smart plug, and smart LED lights, I've setup an "auto-off" after print completion functionality that turns everything off which is also working great.

rsnodgrass commented:
Thanks for adding more sensors!  I was just looking into doing similar, but saw this PR was open.  

One additional sensor I was considering adding was a device_class 'running' `binary_sensor` that is true when the printer is printing. I found basing automations off `*_progress going` to 0 was more problematic than just
having a binary_sensor helper.

Here is my example helper:

```yaml
template:                                                                                               
  - binary_sensor:                                                                                      
      # NOTE: Need a special binary_sensor to track printer running state
      # since the resolution of checking the FlashForge 3 progress sensor
      # sometimes goes from 97/98 percent to 0 (instead of 99/100).
      - name: FlashForge Adventure 3 Running
        device_class: running
        state: >
          {{ states('sensor.flashforge_adventurer_3_progress') > 0 }}
```

This `*_running` `binary_sensor` probably should just be exposed by default through the integration.

@modrzew Look forward to this being merged!

rsnodgrass commented:
@modrzew Just saw your note about not having a FF printer anymore and you don't have time for this anymore. Maybe adding permissions for @f00d4tehg0dz to approve merges/update this repo eases the transition until HACS is update. Then the community has an "owner" that is actively maintaining this in the interim.  THANK YOU FOR CREATING THIS!

f00d4tehg0dz commented:
Great @rsnodgrass  I can go ahead and add that!

@modrzew, whatever you decide, I'm okay with it. You're also welcome to DM me on Discord if you want to go over any specifics or concerns `f00d4tehg0dz`

One thing to note for everyone: I moved from Flashforge stock to Klipper firmware to support the [ERCF ](https://github.com/EtteGit/EnragedRabbitProject)and am finishing the build. However, switching between Stock and Klipper is super easy, so I can maintain this repo just fine. 
I plan on adding a Klipper toggle for the AD5M series to bring in the same data :)

modrzew commented:
@f00d4tehg0dz do you want to merge this PR instead to the master branch of your fork (https://github.com/f00d4tehg0dz/hass-flashforge-adventurer-3)? Once that's done we can swap the integration link in HACS by changing the reference here: https://github.com/hacs/default/blob/master/integration#L632, and you can then fully take over. I'll also then archive this repo.

rsnodgrass commented:
> @modrzew & @f00d4tehg0dz something you may want to consider is moving the repository to an Organization. As an OSS maintainer myself this has been the easiest way to make transitions between maintainers instead of having to deal with future path changes and ensuring that everyone using the software is able to successfully make the switch.

Interesting, what project did you do this for? Would love to see how this worked out. Was it one Organization per project?

timnolte commented:
@rsnodgrass yes, in the case of a couple of my projects each had their own Organization, this has made it easy to manage maintainers without the need to change repositories. This wasn't for HA add-on but for WordPress plugins but the same concept applies.

f00d4tehg0dz commented:
> @f00d4tehg0dz do you want to merge this PR instead to the master branch of your fork (https://github.com/f00d4tehg0dz/hass-flashforge-adventurer-3)? Once that's done we can swap the integration link in HACS by changing the reference here: https://github.com/hacs/default/blob/master/integration#L632, and you can then fully take over. I'll also then archive this repo.

Sounds great! I'll get to it this week. New PC, and Windows install so getting all my ducks in a row atm 😄

f00d4tehg0dz committed:
Merge branch 'modrzew:master' into feature/flashforge-adventurer-5m-pro

modrzew commented:
@f00d4tehg0dz you'll need to ensure that two Github actions are running in your fork:
- HACS action: https://hacs.xyz/docs/publish/action/
- hassfest action: https://developers.home-assistant.io/blog/2020/04/16/hassfest/
Moreover, can you update the `manifest.json` file as well? For context, all of that will be needed before we can update the integration link in HACS: https://github.com/hacs/default/pull/2613

f00d4tehg0dz committed:
Update hacs_action.yml

f00d4tehg0dz commented:
> @f00d4tehg0dz you'll need to ensure that two Github actions are running in your fork:

Based on the Checklist, it's completed.  @modrzew 

- [X] I've read the [publishing documentation](https://hacs.xyz/docs/publish/start).
- [X]  I've added the [HACS action](https://hacs.xyz/docs/publish/action) to my repository.
- [X]  (For integrations only) I've added the [hassfest action](https://developers.home-assistant.io/blog/2020/04/16/hassfest/) to my repository.
- [X]  The actions are passing without any disabled checks in my repository.
- [X]  I've added a link to the action run on my repository below in the links section.
- [X]  I've created a new release of the repository after the validation actions were run successfully.


Links
Link to current release: https://github.com/f00d4tehg0dz/hass-flashforge-adventurer-3/releases/tag/v1.0.8
Link to successful HACS action (without the ignore key): https://github.com/f00d4tehg0dz/hass-flashforge-adventurer-3/actions/workflows/hacs_action.yml
Link to successful hassfest action (if integration): https://github.com/f00d4tehg0dz/hass-flashforge-adventurer-3/actions/workflows/hassfest.yml

timnolte commented:
@mistraller I've not seen any other integration for FlashForge anywhere so what did you actually switch to? Unless you modded your 5M series to run Klipper and use the Klipper mod I haven't seen anything else official in HA or listed in HACS.
